app/forms/delivery_partner_form.rb

```diff
@@ -86,10 +86,10 @@ def lead_providers_validation
   def cohorts_validation
     # Ensure all selected lead providers have at least one selected cohort
     # This is indicated by the presence of a provider relationship for that lead provider
-    lead_provider_ids&.filter(&:present?)&.any? do |lead_provider_id|
-      unless chosen_provider_relationships.pluck(:lead_provider_id).include?(lead_provider_id)
-        errors.add(:provider_relationship_hashes, :blank, message: "Choose at least one cohort for every lead provider")
-      end
+    # Boils down to a set comparison of Arrays of lead_provider_ids
+    # i.e. (A-B) returns all elements in A not included in B, so if there are any, that a validation failure.
+    if (lead_provider_ids&.filter(&:present?)-chosen_provider_relationships.pluck(:lead_provider_id)).any?
+      errors.add(:provider_relationship_hashes, :blank, message: "Choose at least one cohort for every lead provider")
```

Comment:
There are now 4 lines of code comments here, which I doubt anyone will read in full.

I reckon introducing some local variables with expressive names will do more for code maintainability than code comments.
```
selected_lead_provider_ids = lead_provider_ids&.filter(&:present?)
lead_providers_ids_with_cohort = leachosen_provider_relationships.pluck(:lead_provider_id)
if (selected_lead_provider_ids - lead_providers_ids_with_cohort).any?
  ...
```
That said, I think we are spending more time on it than it's worth :P

Reply:
I've extracted the `lead_provider_ids&filter(&:present?)`into a method since it's used in more than one place. The new comment was more to show the rationale behind it, and I'm not sure how valuable the original comment is. My initial thought was to extract into a method with a name close to that comment description. You're right that we're spending a long time over a tiny issue though. Working code replacing broken code was the desire here.